(1) 23.08.28 마크다운 정리규칙 ? 을 정하면 좋겠다 ..
(2) 

!['평가 기준'](ref.png)

### EDA, 탐색적 데이터 분석

> 필요 라이브러리 import

In [57]:
import pandas as pd
import os 
import numpy as np

In [2]:
os.chdir('C:/Users/mlosv/Downloads/23.08.22 데이터분석경진대회')

In [3]:
# CSV 파일 경로 지정
file_name = 'loan_customer.csv'

In [13]:
# CSV 파일 읽기
df = pd.read_csv(file_name)

In [17]:
# 컬럼 정보 미리 붙여 보기 (지주 제공 ppt참고)
column_info = {'SK_ID_CURR':'대출식별번호',
'NAME_CONTRACT_TYPE':'계약유형',
'CODE_GENDER':'성별',
'FLAG_OWN_REALTY':'자가부동산보유여부',
'CNT_CHILDREN':'자녀수',
'AMT_INCOME_TOTAL':'소득금액(USD)',
'AMT_CREDIT':'신용대출금액(USD)',
'NAME_INCOME_TYPE':'소득유형',
'NAME_EDUCATION_TYPE':'학력',
'NAME_FAMILY_STATUS':'결혼상태',
'NAME_HOUSING_TYPE':'주거형태',
'REGION_POPULATION_RELATIVE':'거주지역의표준화된인구',
'DAYS_BIRTH':'대출신청시점에서의일단위나이(기준시점대비상대적인시간)',
'DAYS_EMPLOYED':'대출신청시점에서의고객의근무일수(기준시점대비상대적인시간)',
'DAYS_REGISTRATION':'대출신청이전에고객이등록정보를변경한시점사이의일수(기준시점대비상대적인시간)',
'DAYS_ID_PUBLISH':'대출신청이전에고객이신분증을변경한시점사이의일수(기준시점대비상대적인시간)',
'FLAG_MOBIL':'휴대폰번호등록여부',
'FLAG_EMP_PHONE':'고용주연락처등록여부',
'FLAG_WORK_PHONE':'직장전화번호등록여부',
'FLAG_PHONE':'집전화번호등록여부',
'FLAG_EMAIL':'이메일주소등록여부',
'REGION_RATING_CLIENT':'거주지등급',
'WEEKDAY_APPR_PROCESS_START':'대출신청요일',
'HOUR_APPR_PROCESS_START':'대출신청시간대',
'REG_CITY_NOT_LIVE_CITY':'고객의등록지주소와실거주지일치여부',
'REG_CITY_NOT_WORK_CITY':'고객의등록지주소와직장주소일치여부',
'LIVE_CITY_NOT_WORK_CITY':'고객의실거주지와직장주소일치여부',
'ORGANIZATION_TYPE':'근무회사유형',
'FLAG_DOCUMENT_2':'문서2제공여부',
'FLAG_DOCUMENT_3':'문서3제공여부',
'FLAG_DOCUMENT_4':'문서4제공여부',
'CB_SCORE_1':'신용점수1',
'CB_SCORE_2':'신용점수2',
'CB_SCORE_3':'신용점수3',
'TARGET':'연체여부'}

In [21]:
# 컬럼 정보 수정 
df_v2 = df.copy() 

for col in df_v2.columns : 
    df_v2.rename(columns = {col:column_info[col]}, inplace = True)
    print(col+'컬럼의 이름을 '+column_info[col]+'로 수정했습니다 ')

SK_ID_CURR컬럼의 이름을 대출식별번호로 수정했습니다 
NAME_CONTRACT_TYPE컬럼의 이름을 계약유형로 수정했습니다 
CODE_GENDER컬럼의 이름을 성별로 수정했습니다 
FLAG_OWN_REALTY컬럼의 이름을 자가부동산보유여부로 수정했습니다 
CNT_CHILDREN컬럼의 이름을 자녀수로 수정했습니다 
AMT_INCOME_TOTAL컬럼의 이름을 소득금액(USD)로 수정했습니다 
AMT_CREDIT컬럼의 이름을 신용대출금액(USD)로 수정했습니다 
NAME_INCOME_TYPE컬럼의 이름을 소득유형로 수정했습니다 
NAME_EDUCATION_TYPE컬럼의 이름을 학력로 수정했습니다 
NAME_FAMILY_STATUS컬럼의 이름을 결혼상태로 수정했습니다 
NAME_HOUSING_TYPE컬럼의 이름을 주거형태로 수정했습니다 
REGION_POPULATION_RELATIVE컬럼의 이름을 거주지역의표준화된인구로 수정했습니다 
DAYS_BIRTH컬럼의 이름을 대출신청시점에서의일단위나이(기준시점대비상대적인시간)로 수정했습니다 
DAYS_EMPLOYED컬럼의 이름을 대출신청시점에서의고객의근무일수(기준시점대비상대적인시간)로 수정했습니다 
DAYS_REGISTRATION컬럼의 이름을 대출신청이전에고객이등록정보를변경한시점사이의일수(기준시점대비상대적인시간)로 수정했습니다 
DAYS_ID_PUBLISH컬럼의 이름을 대출신청이전에고객이신분증을변경한시점사이의일수(기준시점대비상대적인시간)로 수정했습니다 
FLAG_MOBIL컬럼의 이름을 휴대폰번호등록여부로 수정했습니다 
FLAG_EMP_PHONE컬럼의 이름을 고용주연락처등록여부로 수정했습니다 
FLAG_WORK_PHONE컬럼의 이름을 직장전화번호등록여부로 수정했습니다 
FLAG_PHONE컬럼의 이름을 집전화번호등록여부로 수정했습니다 
FLAG_EMAIL컬럼의 이름을 이메일주소등록여부로 수정했습니다 
REGION_RATING_CLIENT컬럼의 이름을 거주지등급로 수정했습니다 
WEEKDAY_APPR_PROCESS_START

> 전체 데이터 구조 요약해서 보기

In [30]:
df.head()

SK_ID_CURR NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_REALTY  CNT_CHILDREN  \
0      100002         Cash loans           M               Y             0   
1      100004    Revolving loans         NaN             NaN             0   
2      100020         Cash loans           M               N             0   
3      100029         Cash loans           M               N             2   
4      100048         Cash loans         NaN               Y             0   

   AMT_INCOME_TOTAL  AMT_CREDIT NAME_INCOME_TYPE  \
0          202500.0         NaN          Working   
1           67500.0    135000.0          Working   
2          108000.0    509602.5          Working   
3          135000.0         NaN          Working   
4          202500.0    604152.0          Working   

             NAME_EDUCATION_TYPE    NAME_FAMILY_STATUS  ...  \
0  Secondary / secondary special  Single / not married  ...   
1  Secondary / secondary special  Single / not married  ...   
2  Secondary / secondary special               Married  ...   
3  Secondary / secondary special               Married  ...   
4  Secondary / secondary special               Married  ...   

  REG_CITY_NOT_WORK_CITY  LIVE_CITY_NOT_WORK_CITY       ORGANIZATION_TYPE  \
0                      0                        0  Business Entity Type 3   
1                      0                        0              Government   
2                      1                        0              Government   
3                      1                        1  Business Entity Type 3   
4                      1                        1        Industry: type 1   

   FLAG_DOCUMENT_2  FLAG_DOCUMENT_3  FLAG_DOCUMENT_4  CB_SCORE_1  CB_SCORE_2  \
0                0                1                0         NaN         NaN   
1                0                0                0   26.499494   57.600542   
2                0                1                0         NaN   63.496369   
3                0                1                0   82.376245   45.996152   
4                0                1                0   66.199985         NaN   

   CB_SCORE_3  TARGET  
0   58.349224       1  
1   46.433666       0  
2         NaN       0  
3   68.694333       0  
4         NaN       0  

[5 rows x 35 columns]

*1.1 Check Point → 대출식별번호에 중복 없나*

In [15]:
# 대출식별 기준으로 그룹화하고 counting
grouped = df.groupby('SK_ID_CURR').size()

# counting이 1 넘어간 것들만 필터링
filtered_grouped = grouped[grouped > 1]

print(filtered_grouped)

Series([], dtype: int64)


*1.1 결과→ 대출일련번호가 unique key. 그럼 고객 구분은(고객:대출 = 1:n 관계)은 어떻게 하나?*

*1.2 Check Point → 수치/범주형 변수들 타입이 바르게 들어갔나 ?*

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24000 entries, 0 to 23999
Data columns (total 35 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   SK_ID_CURR                  24000 non-null  int64  
 1   NAME_CONTRACT_TYPE          24000 non-null  object 
 2   CODE_GENDER                 16906 non-null  object 
 3   FLAG_OWN_REALTY             14429 non-null  object 
 4   CNT_CHILDREN                24000 non-null  int64  
 5   AMT_INCOME_TOTAL            24000 non-null  float64
 6   AMT_CREDIT                  16700 non-null  float64
 7   NAME_INCOME_TYPE            24000 non-null  object 
 8   NAME_EDUCATION_TYPE         24000 non-null  object 
 9   NAME_FAMILY_STATUS          24000 non-null  object 
 10  NAME_HOUSING_TYPE           24000 non-null  object 
 11  REGION_POPULATION_RELATIVE  16851 non-null  float64
 12  DAYS_BIRTH                  19251 non-null  float64
 13  DAYS_EMPLOYED               168

*1.2 결과→ 수치형 변수는 float/int, 범주형 변수는 object로 잘 들어가있음*

*1.3 Check Point → 결측값이 어떻게 들어가있는지 모르니, 결측치 유효값들 살펴보기 위해 value_coutns*

In [82]:
for column in df.columns:
    value_counts = df[column].value_counts(dropna = False)
    print(f"Value counts for {column} {column_info[column]}:\n{value_counts}\n")

Value counts for SK_ID_CURR 대출식별번호:
100002    1
337229    1
337317    1
337293    1
337288    1
         ..
218660    1
218625    1
218609    1
218594    1
456228    1
Name: SK_ID_CURR, Length: 24000, dtype: int64

Value counts for NAME_CONTRACT_TYPE 계약유형:
Cash loans         21733
Revolving loans     2267
Name: NAME_CONTRACT_TYPE, dtype: int64

Value counts for CODE_GENDER 성별:
F      11211
NaN     7094
M       5694
XNA        1
Name: CODE_GENDER, dtype: int64

Value counts for FLAG_OWN_REALTY 자가부동산보유여부:
Y      9939
NaN    9571
N      4490
Name: FLAG_OWN_REALTY, dtype: int64

Value counts for CNT_CHILDREN 자녀수:
0    16859
1     4703
2     2046
3      341
4       41
5        9
7        1
Name: CNT_CHILDREN, dtype: int64

Value counts for AMT_INCOME_TOTAL 소득금액(USD):
135000.0    2779
112500.0    2422
157500.0    2052
180000.0    1860
90000.0     1785
            ... 
68400.0        1
118800.0       1
183375.0       1
113008.5       1
120784.5       1
Name: AMT_INCOME_TOTAL, Length: 494, dty

> 결측치 종류 : 'NaN', 'nan', 'XNA', 'Unknown' ▶ np.nan로 통일 

In [83]:
adj_df = df.copy()

In [84]:
adj_df = adj_df.astype(str)
adj_df = adj_df.replace(['NaN', 'nan', 'XNA', 'Unknown'], np.nan)

In [85]:
adj_df = adj_df.astype(df.dtypes)

In [88]:
for column in adj_df.columns:
    value_counts = adj_df[column].value_counts(dropna = False)
    print(f"Value counts for {column} {column_info[column]}:\n{value_counts}\n")

Value counts for SK_ID_CURR 대출식별번호:
100002    1
337229    1
337317    1
337293    1
337288    1
         ..
218660    1
218625    1
218609    1
218594    1
456228    1
Name: SK_ID_CURR, Length: 24000, dtype: int64

Value counts for NAME_CONTRACT_TYPE 계약유형:
Cash loans         21733
Revolving loans     2267
Name: NAME_CONTRACT_TYPE, dtype: int64

Value counts for CODE_GENDER 성별:
F      11211
NaN     7095
M       5694
Name: CODE_GENDER, dtype: int64

Value counts for FLAG_OWN_REALTY 자가부동산보유여부:
Y      9939
NaN    9571
N      4490
Name: FLAG_OWN_REALTY, dtype: int64

Value counts for CNT_CHILDREN 자녀수:
0    16859
1     4703
2     2046
3      341
4       41
5        9
7        1
Name: CNT_CHILDREN, dtype: int64

Value counts for AMT_INCOME_TOTAL 소득금액(USD):
135000.0    2779
112500.0    2422
157500.0    2052
180000.0    1860
90000.0     1785
            ... 
68400.0        1
118800.0       1
183375.0       1
113008.5       1
120784.5       1
Name: AMT_INCOME_TOTAL, Length: 494, dtype: int64

Va

*1.3 결과 → 결측치 표기 np.nan로 통일*

*1.4 Check Point → 범주형 변수들(object객체)의 유효값 체크*

In [89]:
# object 타입인 열들만 선택
object_columns = adj_df.select_dtypes(include=['object'])

# 각 열의 값의 빈도를 value_counts()로 계산
for column in object_columns:
    value_counts = object_columns[column].value_counts(dropna = False)
    print(f"Value counts for {column} {column_info[column]}:\n{value_counts}\n")

Value counts for NAME_CONTRACT_TYPE 계약유형:
Cash loans         21733
Revolving loans     2267
Name: NAME_CONTRACT_TYPE, dtype: int64

Value counts for CODE_GENDER 성별:
F      11211
NaN     7095
M       5694
Name: CODE_GENDER, dtype: int64

Value counts for FLAG_OWN_REALTY 자가부동산보유여부:
Y      9939
NaN    9571
N      4490
Name: FLAG_OWN_REALTY, dtype: int64

Value counts for NAME_INCOME_TYPE 소득유형:
Working                 12460
Commercial associate     5527
Pensioner                4294
State servant            1714
Student                     2
Maternity leave             2
Businessman                 1
Name: NAME_INCOME_TYPE, dtype: int64

Value counts for NAME_EDUCATION_TYPE 학력:
Secondary / secondary special    17054
Higher education                  5857
Incomplete higher                  796
Lower secondary                    278
Academic degree                     15
Name: NAME_EDUCATION_TYPE, dtype: int64

Value counts for NAME_FAMILY_STATUS 결혼상태:
Married                 15275
Single / 

*1.4 결과 → (1) 각 범주형 변수들별로 크게 묶을 수 있는 것들은 대분류 변수들 생성  
(2) 성별 / 자가부동산보유여부에 결측값이 생각보다 많음 ▶ 어떻게 처리할지 고민 필요*

< 유효값 정보 참고 > 

1. NAME_INCOME_TYPE 소득유형
Working - 근로자 : 12460  
Commercial associate - 상업 공동주택 : 5527  
Pensioner - 연금 수령자 : 4294  
State servant - 공무원 : 1714  
Student - 학생 : 2  
Maternity leave - 육아 휴가 : 2     
Businessman - 사업가 : 1   


2. NAME_EDUCATION_TYPE 학력  
Secondary / secondary special - 중등 교육 / 중등 특별 교육: 17054  
Higher education - 고등 교육: 5857  
Incomplete higher - 미완료된 고등 교육: 796  
Lower secondary - 초등 교육: 278  
Academic degree - 학위: 15  


3. NAME_FAMILY_STATUS 결혼상태
Married - 기혼: 15275  
Single / not married - 미혼: 3557  
Civil marriage - 민사결혼: 2336  
Separated - 별거: 1527  
Widow - 사별: 1304  
Unknown - 알 수 없음: 1  


4. NAME_HOUSING_TYPE 주거형태
House / apartment - 주택 / 아파트: 21213  
With parents - 부모님과 함께: 1206  
Municipal apartment - 시립 아파트: 911  
Rented apartment - 임대 아파트: 391  
Office apartment - 사무 아파트: 203  
Co-op apartment - 주택조합 아파트: 76  


5. ORGANIZATION_TYPE 근무회사유형
Business Entity Type 3 - 사업 엔터티 유형 3: 5311    
XNA - XNA: 4294  
Self-employed - 자영업: 3002  
Other - 기타: 1285  
Medicine - 의학: 855  
Government - 정부: 790  
Business Entity Type 2 - 사업 엔터티 유형 2: 788  
School - 학교: 685  
Trade: type 7 - 거래: 유형 7: 636  
Kindergarten - 유치원: 530  
Construction - 건설: 491  
Business Entity Type 1 - 사업 엔터티 유형 1: 474  
Transport: type 4 - 운송: 유형 4: 429  
Industry: type 9 - 산업: 유형 9: 286  
Trade: type 3 - 거래: 유형 3: 260    
Industry: type 3 - 산업: 유형 3: 256  
Security - 보안: 248  
Bank - 은행: 236  
Housing - 주택: 232  
Agriculture - 농업: 202  
Military - 군대: 201  
Industry: type 11 - 산업: 유형 11: 192  
Police - 경찰: 182  
Postal - 우체국: 176  
Transport: type 2 - 운송: 유형 2: 170  
Trade: type 2 - 거래: 유형 2: 166    
Security Ministries - 보안부: 139  
Services - 서비스: 136  
Restaurant - 레스토랑: 122  
Industry: type 7 - 산업: 유형 7: 111  
Transport: type 3 - 운송: 유형 3: 99  
University - 대학교: 97  
Hotel - 호텔: 91  
Industry: type 1 - 산업: 유형 1: 89  
Electricity - 전기: 88  
Industry: type 4 - 산업: 유형 4: 75  
Emergency - 비상: 61  
Trade: type 6 - 거래: 유형 6: 59  
Telecom - 통신: 54  
Industry: type 5 - 산업: 유형 5: 49  
Insurance - 보험: 48  
Advertising - 광고: 44  
Industry: type 2 - 산업: 유형 2: 34  
Trade: type 1 - 거래: 유형 1: 29  
Legal Services - 법률 서비스: 27  
Industry: type 12 - 산업: 유형 12: 26  
Culture - 문화: 25  
Realtor - 부동산 중개인: 25  
Mobile - 모바일: 23  
Cleaning - 청소: 20  
Transport: type 1 - 운송: 유형 1: 19  
Industry: type 6 - 산업: 유형 6: 7  
Trade: type 4 - 거래: 유형 4: 7  
Industry: type 13 - 산업: 유형 13: 5  
Religion - 종교: 5  
Industry: type 10 - 산업: 유형 10: 4  
Trade: type 5 - 거래: 유형 5: 4  
Industry: type 8 - 산업: 유형 8: 1  

*1.4.대분류 변수 생성*  
**ORGANIZATION_TYPE 근무회사유형**

In [125]:
adj_df['ADJ_ORGANIZATION_TYPE'] = adj_df['ORGANIZATION_TYPE'].copy()

In [126]:
adj_df['ADJ_ORGANIZATION_TYPE'] = adj_df['ADJ_ORGANIZATION_TYPE'].str.replace(r'^Trade.*', 'Trade', regex = True)
adj_df['ADJ_ORGANIZATION_TYPE'] = adj_df['ADJ_ORGANIZATION_TYPE'].str.replace('^Industry.*', 'Industry', regex = True)
adj_df['ADJ_ORGANIZATION_TYPE'] = adj_df['ADJ_ORGANIZATION_TYPE'].str.replace('^Transport.*', 'Transport', regex = True)
adj_df['ADJ_ORGANIZATION_TYPE'] = adj_df['ADJ_ORGANIZATION_TYPE'].str.replace('^Business Entity.*', 'Business Entity', regex = True)

In [127]:
adj_df['ADJ_ORGANIZATION_TYPE'].value_counts(dropna=False)

Business Entity        6573
NaN                    4294
Self-employed          3002
Other                  1285
Trade                  1161
Industry               1135
Medicine                855
Government              790
Transport               717
School                  685
Kindergarten            530
Construction            491
Security                248
Bank                    236
Housing                 232
Agriculture             202
Military                201
Police                  182
Postal                  176
Security Ministries     139
Services                136
Restaurant              122
University               97
Hotel                    91
Electricity              88
Emergency                61
Telecom                  54
Insurance                48
Advertising              44
Legal Services           27
Culture                  25
Realtor                  25
Mobile                   23
Cleaning                 20
Religion                  5
Name: ADJ_ORGANIZATI

In [134]:
# 직업군을 그룹핑한 딕셔너리 생성
organization_type_dict = {
    'Business and Self-employed': ['Business Entity', 'Self-employed', 'Trade'],
    'Industry and Construction': ['Industry', 'Construction'],
    'Medicine and Healthcare': ['Medicine'],
    'Government and Security': ['Government', 'Security', 'Security Ministries', 'Military', 'Police'],
    'Education': ['School', 'Kindergarten', 'University'],
    'Housing and Real Estate': ['Housing', 'Bank', 'Real Estate'],
    'Logistics and Services': ['Transport', 'Postal', 'Services'],
    'Hospitality': ['Restaurant', 'Hotel'],
    'Utilities and Telecom': ['Electricity', 'Telecom'],
    'Others': ['Other', 'Agriculture', 'Emergency', 'Insurance', 'Advertising', 'Legal Services', 'Culture', 'Realtor', 'Mobile', 'Cleaning', 'Religion']
}

In [138]:
# {
#     '사업 및 자영업': ['사업 엔터티', '자영업', '거래'],
#     '산업 및 건설': ['산업', '건설'],
#     '의학 및 의료': ['의학'],
#     '정부 및 보안': ['정부', '보안', '보안 기관', '군대', '경찰'],
#     '교육': ['학교', '유치원', '대학교'],
#     '주택 및 부동산': ['주택', '은행', '부동산'],
#     '물류 및 서비스': ['운송', '우편', '서비스'],
#     '식음료업': ['레스토랑', '호텔'],
#     '공공 및 통신사업': ['전기', '통신'],
#     '기타': ['기타', '농업', '비상', '보험', '광고', '법률 서비스', '문화', '부동산 중개인', '휴대폰', '청소', '종교']
# }

In [135]:
for category, occupations in group_mapping.items():
    adj_df.loc[adj_df['ADJ_ORGANIZATION_TYPE'].isin(occupations), 'ADJ_ORGANIZATION_TYPE'] = category

In [136]:
adj_df['ADJ_ORGANIZATION_TYPE'].value_counts(dropna=False)

Business and Self-employed    10736
NaN                            4294
Others                         1765
Industry and Construction      1626
Government and Security        1560
Education                      1312
Logistics and Services         1029
Medicine and Healthcare         855
Housing and Real Estate         468
Hospitality                     213
Utilities and Telecom           142
Name: ADJ_ORGANIZATION_TYPE, dtype: int64

In [ ]:

# 데이터의 처음 5개 행 출력
print("처음 5개 행 데이터:")
print(data.head())

# 데이터의 기본 정보 출력
print("\n데이터 기본 정보:")
print(data.info())

# 수치형 변수의 기술 통계 출력
print("\n수치형 변수의 기술 통계:")
print(data.describe())

# 범주형 변수의 값 세기
categorical_columns = data.select_dtypes(include=['object']).columns
print("\n범주형 변수의 값 세기:")
for col in categorical_columns:
    print(data[col].value_counts())